proxy/modules/native_module_host/src/native_module_host.c

```diff
@@ -0,0 +1,418 @@
+	{
+		/*Codes_SRS_NATIVEMODULEHOST_17_008: [ NativeModuleHost_Create shall return NULL if broker is NULL. ]*/
+		/*Codes_SRS_NATIVEMODULEHOST_17_009: [ NativeModuleHost_Create shall return NULL if configuration does not contain valid JSON. ]*/
+		LogError("broker or configuration is NULL, both are required");
+		result = NULL;
+	}
+	else
+	{
+		/*Codes_SRS_NATIVEMODULEHOST_17_010: [ NativeModuleHost_Create shall intialize the Module_Loader. ]*/
+		if (ModuleLoader_Initialize() != MODULE_LOADER_SUCCESS)
+		{
+			/*Codes_SRS_NATIVEMODULEHOST_17_026: [ If any step above fails, then NativeModuleHost_Create shall free all resources allocated and return NULL. ]*/
+			LogError("ModuleLoader_Initialize failed");
+			result = NULL;
+		}
+		else
+		{
+			/*Codes_SRS_NATIVEMODULEHOST_17_011: [ NativeModuleHost_Create shall parse the configuration JSON. ]*/
+			char * outprocess_module_args = (char *)configuration;
+			JSON_Value *root_value = json_parse_string(outprocess_module_args);
+			JSON_Object * module_host_args;
+			if ((root_value == NULL) ||
+				((module_host_args = json_value_get_object(root_value)) == NULL))
+			{
+				if (root_value != NULL)
+				{
+					json_value_free(root_value);
+				}
+				LogError("NativeModuleHost_Create could not parse arguments as JSON");
+				result = NULL;
+			}
+			else
+			{
+				/*Codes_SRS_NATIVEMODULEHOST_17_035: [ If the "outprocess.loaders" array exists in the configuration JSON, NativeModuleHost_Create shall initialize the Module_Loader from this array. ]*/
+				JSON_Value * loaders_array = json_object_get_value(module_host_args, OOP_MODULE_LOADERS_ARRAY_KEY);
+				if ((loaders_array != NULL) &&
+					(ModuleLoader_InitializeFromJson(loaders_array) != MODULE_LOADER_SUCCESS))
+				{
+					LogError("NativeModuleHost_Create could not extract loaders array from module arguments.");
+					result = NULL;
+				}
+				else
+				{
+					/*Codes_SRS_NATIVEMODULEHOST_17_012: [ NativeModuleHost_Create shall get the "outprocess.loader" object from the configuration JSON. ]*/
+					JSON_Object * loader_args = json_object_get_object(module_host_args, OOP_MODULE_LOADER_KEY);
+					if (loader_args == NULL)
+					{
+						LogError("NativeModuleHost_Create could not get loader arguments.");
+						result = NULL;
+					}
+					else
+					{
+						GATEWAY_MODULE_LOADER_INFO loader_info;
+						if (parse_loader(loader_args, &loader_info) != 0)
+						{
+							/*Codes_SRS_NATIVEMODULEHOST_17_026: [ If any step above fails, then NativeModuleHost_Create shall free all resources allocated and return NULL. ]*/
+							LogError("NativeModuleHost_Create could not extract loader information from loader arguments.");
+							result = NULL;
+						}
+						else
+						{
+							// Have loader and entrypoint now, get module.
+							result = (MODULE_HOST*)malloc(sizeof(MODULE_HOST));
+							/*Codes_SRS_NATIVEMODULEHOST_17_018: [ NativeModuleHost_Create shall get the "module.args" object from the configuration JSON. ]*/
```

Comment:
So, if it faile module_create is going to fail, but you will try to free a failed memory... So....

---
In reply to: [106247605](https://github.com/Azure/azure-iot-gateway-sdk/pull/160#discussion_r106247605) [](ancestors = 106247605)